File: ros/angel_utils/src/rate_tracker.cpp
#include "../include/angel_utils/rate_tracker.hpp"

#include <chrono>
#include <numeric>
#include <vector>

namespace angel_utils {

namespace {

/// Type of clock to use for tick rate tracking.
using rt_clock_t = std::chrono::high_resolution_clock;

} // namespace

// ----------------------------------------------------------------------------
struct RateTracker::PImpl
{
  //
  // Properties
  //

  /// The measurement window size for rate smoothing.
  size_t m_window_size = 10;

  /// If we have not received a tick yet.
  bool m_first_tick = true;

  /// The last a tick occurred.
  std::chrono::time_point< rt_clock_t > m_last_measure_time;

  /// Circular buffer of time measurements between ticks.
  std::vector< double > m_time_vec;

  /// When the m_time_vec circular buffer is full, this is the next index to
  /// write to.
  size_t m_time_vec_i = 0;

  //
  // Methods
  //

  /// If we have time measurements to act on or not.
  bool
  has_measurements() const
  {
    return not m_time_vec.empty();
  }
};

// ----------------------------------------------------------------------------
RateTracker
::RateTracker( size_t window_size )
  : impl( std::make_unique< PImpl >() )
{
  impl->m_window_size = window_size;
}

// ----------------------------------------------------------------------------
RateTracker::~RateTracker() = default;

// ----------------------------------------------------------------------------
void
RateTracker
::tick()
{
  auto pub_time = rt_clock_t::now();
  if( impl->m_first_tick )
  {
    impl->m_first_tick = false;
  }
  else
  {
    double time_since_last_tick = std::chrono::duration< double >(
      pub_time - impl->m_last_measure_time
      ).count();
    // Insert time measurement appropriately into window.
    if( impl->m_time_vec.size() < impl->m_window_size )
    {
      // add measurement
      impl->m_time_vec.push_back( time_since_last_tick );
    }
    else
    {
      // m_frame_time_vec is full, so now we start rotating new-measurement
      // insertion
      impl->m_time_vec[ impl->m_time_vec_i ] = time_since_last_tick;
      impl->m_time_vec_i = ( impl->m_time_vec_i + 1 ) %
                           impl->m_time_vec.size();
    }
  }
  impl->m_last_measure_time = pub_time;
}

// ----------------------------------------------------------------------------
double
RateTracker
::get_delta_avg()
{
  double avg_time = -1;
  if( impl->has_measurements() )
  {
    double window_total = std::accumulate( impl->m_time_vec.begin(),
                                           impl->m_time_vec.end(), 0.0 );
    avg_time = window_total / static_cast< double >( impl->m_time_vec.size() );
  }
  return avg_time;
}

// ----------------------------------------------------------------------------
double
RateTracker
::get_rate_avg()
{
  double avg_rate = -1;
  if( impl->has_measurements() )
  {
    auto avg_time = get_delta_avg();
    avg_rate = 1.0 / avg_time;
  }
  return avg_rate;
}

} // namespace angel_utils

File: ros/angel_utils/python/angel_utils/rate_tracker.py
import time


class RateTracker:
    """
    Keep track of the rate at which something is cycling.

    This class is currently *not* thread-safe.

    The C++ sibling of this utility is located [here (impl)](ros/angel_utils/src/rate_tracker.cpp)
    and [here (header)](ros/angel_utils/include/angel_utils/rate_tracker.hpp).
    Changes to this API and implementation should be reflected there.

    :param window_size: Number of tick delta's to retain to compute moving
        averages.
    """

    def __init__(self, window_size: int = 10):
        self._window_size = window_size
        self._first_tick = True
        self._last_measure_time = None
        self._time_vec = []
        self._time_vec_i = 0

    def _has_measurements(self) -> bool:
        """ If we have any time measurements to act on or not. """
        return self._time_vec.size() > 0

    def tick(self) -> None:
        """
        Perform a measurement of time since the last tick.
        """
        now = time.monotonic()
        if self._first_tick:
            self._first_tick = False
        else:
            time_since_last_tick = now - self._last_measure_time
            # Insert time measurement appropriately into the window
            if len(self._time_vec) < self._window_size:
                self._time_vec.append(time_since_last_tick)
            else:
                # _time_vec is full, so now we start rotating new measurement
                # insertion. _time_vec_i start's at 0 when we first enter here,
                # which would overwrite the oldest time measurement at the
                # beginning of the vector.
                self._time_vec[self._time_vec_i] = time_since_last_tick
                self._time_vec_i = (self._time_vec_i + 1) % len(self._time_vec)
        self._last_measure_time = now

    def get_delta_avg(self) -> float:
        """
        Get the average time delta between ticks within our window of
        measurement.

        If there have been no measurements taken yet (by calling `tick()`) then
        a `-1` value is returned.

        :return: Average time in seconds between tick measurements.
        """
        avg_time = -1
        if len(self._time_vec) > 0:
            window_total = sum(self._time_vec)
            avg_time = window_total / len(self._time_vec)
        return avg_time

    def get_rate_avg(self) -> float:
        """
        Get the average tick rate from the window of measurements.

        If there have been no measurements taken yet (by calling `tick()`) then
        a `-1` value is returned.

        :return: Average rate in Hz between tick measurements.
        """
        avg_rate = -1
        if len(self._time_vec) > 0:
            avg_rate = 1.0 / self.get_delta_avg()
        return avg_rate
